File: challenge_8.cpp
//Challenge 8 

#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <string>

using namespace std;

// Function to convert a hex string to a vector of bytes
vector<uint8_t> hexToBytes(const string& hex) {
    vector<uint8_t> bytes;
    for (size_t i = 0; i < hex.length(); i += 2) {
        string byteString = hex.substr(i, 2);
        uint8_t byte = static_cast<uint8_t>(stoi(byteString, nullptr, 16));
        bytes.push_back(byte);
    }
    return bytes;
}

// Function to detect if a ciphertext has been encrypted using ECB mode
bool hasRepeatedBlocks(const vector<uint8_t>& ciphertext, size_t blockSize) {
    set<vector<uint8_t>> uniqueBlocks;
    for (size_t i = 0; i < ciphertext.size(); i += blockSize) {
        vector<uint8_t> block(ciphertext.begin() + i, ciphertext.begin() + i + blockSize);
        if (uniqueBlocks.find(block) != uniqueBlocks.end()) {
            return true; // Repeated block found
        }
        uniqueBlocks.insert(block);
    }
    return false;
}

int main() {
    ifstream file("Cryptopals8.txt");
    vector<string> ciphertexts;
    string line;

    // Read ciphertexts from the file
    if (file.is_open()) {
        while (getline(file, line)) {
            if (!line.empty()) {
                ciphertexts.push_back(line);
            }
        }
        file.close();
    }
    else {
        cerr << "Unable to open file" << endl;
        return 1;
    }

    size_t blockSize = 16; // AES block size (16 bytes)

    // Detect ECB-encrypted ciphertext
    for (size_t i = 0; i < ciphertexts.size(); ++i) {
        vector<uint8_t> ciphertextBytes = hexToBytes(ciphertexts[i]);
        if (hasRepeatedBlocks(ciphertextBytes, blockSize)) {
            cout << "ECB-encrypted ciphertext found at index " << i << endl;
            cout << "Ciphertext: " << ciphertexts[i] << endl;
        }
    }

    return 0;
}

File: challenge8.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class CryptopalsChallenge8
{
    // Function to convert a hex string to a byte array
    static byte[] HexToBytes(string hex)
    {
        int length = hex.Length / 2;
        byte[] bytes = new byte[length];
        for (int i = 0; i < length; i++)
        {
            bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        }
        return bytes;
    }

    // Function to detect if a ciphertext has been encrypted using ECB mode
    static bool HasRepeatedBlocks(byte[] ciphertext, int blockSize)
    {
        HashSet<string> uniqueBlocks = new HashSet<string>();
        for (int i = 0; i < ciphertext.Length; i += blockSize)
        {
            byte[] block = ciphertext.Skip(i).Take(blockSize).ToArray();
            string blockString = BitConverter.ToString(block);
            if (uniqueBlocks.Contains(blockString))
            {
                return true; // Repeated block found
            }
            uniqueBlocks.Add(blockString);
        }
        return false;
    }

    static void Main()
    {
        string filePath = "challenge8.txt";
        List<string> ciphertexts = new List<string>();

        // Read ciphertexts from the file
        try
        {
            var lines = File.ReadAllLines(filePath);
            ciphertexts.AddRange(lines.Where(line => !string.IsNullOrEmpty(line)));
        }
        catch (Exception)
        {
            Console.WriteLine("Unable to open file");
            return;
        }

        int blockSize = 16; // AES block size (16 bytes)

        // Detect ECB-encrypted ciphertext
        for (int i = 0; i < ciphertexts.Count; i++)
        {
            byte[] ciphertextBytes = HexToBytes(ciphertexts[i]);
            if (HasRepeatedBlocks(ciphertextBytes, blockSize))
            {
                Console.WriteLine($"ECB-encrypted ciphertext found at index {i}");
                Console.WriteLine($"Ciphertext: {ciphertexts[i]}");
            }
        }
    }
}
